File: apps/artifacts/api/python/node/tests/node_test.py
import unittest
import typing
import json

from apps.artifacts.api.python.common import NodePublishNoRemote
from apps.artifacts.api.python.node.node import Node
from bzd.http.client_mock import HttpClientMock, HttpResponseMock


class TestRun(unittest.TestCase):
	def testPublish(self) -> None:
		self.calledCounter = 0

		def callback(method: str, url: str, body: typing.Optional[bytes], **kwargs: typing.Any) -> None:
			self.calledCounter += 1
			self.assertEqual(method, "POST")
			assert body is not None
			data = json.loads(body.decode())
			# Navigate: data["data"][0] = [subKey, pairs]; pairs[0] = [ts, value]; value is [1]
			self.assertEqual(data["data"][0][1][0][1], "hello")
			self.assertIn("timestamp", data)

		node = Node(uid="testuid", httpClient=HttpClientMock(callback=callback))
		node.publish("hello")
		self.assertEqual(self.calledCounter, 1)

	def testPublishBulk(self) -> None:
		self.calledCounter = 0

		def callback(method: str, url: str, body: typing.Optional[bytes], **kwargs: typing.Any) -> None:
			self.calledCounter += 1
			assert body is not None
			data = json.loads(body.decode())
			self.assertEqual(
				data["data"],
				[[[], [[1, {"hello": "world1"}]]], [[], [[2, {"hello": "world2"}]]]],
			)
			self.assertIn("timestamp", data)

		node = Node(uid="testuid", httpClient=HttpClientMock(callback=callback))
		with node.publishBulk() as accessor:
			accessor(timestampMs=1, value={"hello": "world1"})
			accessor(timestampMs=2, value={"hello": "world2"})
		self.assertEqual(self.calledCounter, 1)

	def testPublishBulkFixedTimestamp(self) -> None:
		self.calledCounter = 0

		def callback(method: str, url: str, body: typing.Optional[bytes], **kwargs: typing.Any) -> None:
			self.calledCounter += 1
			assert body is not None
			data = json.loads(body.decode())
			self.assertEqual(
				data,
				{"data": [[[], [[1, {"hello": "world1"}]]], [[], [[2, {"hello": "world2"}]]]]},
			)
			self.assertNotIn("timestamp", data)

		node = Node(uid="testuid", httpClient=HttpClientMock(callback=callback))
		with node.publishBulk(isClientTimestamp=False) as accessor:
			accessor(timestampMs=1, value={"hello": "world1"})
			accessor(timestampMs=2, value={"hello": "world2"})
		self.assertEqual(self.calledCounter, 1)

	def testPublishMultiNodes(self) -> None:
		self.calledCounter = 0

		def callback(method: str, url: str, body: typing.Optional[bytes], **kwargs: typing.Any) -> None:
			self.calledCounter += 1
			self.assertEqual(method, "POST")
			self.assertEqual(url, "http://localhost:8081/x/nodes/?bulk=1")
			assert body is not None
			data = json.loads(body.decode())
			self.assertEqual(data["data"]["hello"][0][1][0][1], "1")
			self.assertEqual(data["data"]["world"][0][1][0][1], "2")
			self.assertIn("timestamp", data)

		node = Node(uid="testuid", httpClient=HttpClientMock(callback=callback))
		node.publishMultiNodes(data={"hello": "1", "world": "2"})
		self.assertEqual(self.calledCounter, 1)

	def testPublishNoServer(self) -> None:

		def callback(**kwargs: typing.Any) -> None:
			raise Exception("Unreachable")

		node = Node(uid="testuid", httpClient=HttpClientMock(callback=callback))
		self.assertRaises(NodePublishNoRemote, lambda: node.publish(1))

	def testPublishNoServerWithBuffer(self) -> None:
		self.calledCounter = 0
		values: typing.List[typing.Any] = []
		self.callbackRaise = True

		def callback(method: str, url: str, body: typing.Optional[bytes], **kwargs: typing.Any) -> None:
			if self.callbackRaise:
				raise Exception("Unreachable")
			self.calledCounter += 1
			assert body is not None
			data = json.loads(body.decode())
			values.append(data["data"][0][1][0][1])

		node = Node(uid="testuid", httpClient=HttpClientMock(callback=callback), maxBufferSize=2)
		self.assertRaises(NodePublishNoRemote, lambda: node.publish(2))
		self.assertEqual(len(node.buffer), 1)
		self.assertRaises(NodePublishNoRemote, lambda: node.publish(3))
		self.assertEqual(len(node.buffer), 2)
		self.assertRaises(NodePublishNoRemote, lambda: node.publish(4))
		# The buffer is capped at maxBufferSize, the oldest entry is dropped.
		self.assertEqual(len(node.buffer), 2)
		self.assertEqual(self.calledCounter, 0)

		self.callbackRaise = False
		node.publish(5)
		self.assertEqual(self.calledCounter, 2)
		self.assertEqual(values, [4, 5])

	def testPublishErrorNextRemote(self) -> None:
		self.calledCounter = 0

		def callback(method: str, url: str, body: typing.Optional[bytes], **kwargs: typing.Any) -> None:
			if url.startswith("http://remote1"):
				raise Exception("Unreachable")
			self.calledCounter += 1

		node = Node(uid="testuid", remotes=["http://remote1", "http://remote2"], httpClient=HttpClientMock(callback=callback))
		node.publish("hello")
		self.assertEqual(self.calledCounter, 1)

	def testPublishErrorAllRemotes(self) -> None:

		def callback(**kwargs: typing.Any) -> None:
			raise Exception("Unreachable")

		node = Node(uid="testuid", remotes=["http://remote1", "http://remote2"], httpClient=HttpClientMock(callback=callback))
		self.assertRaises(NodePublishNoRemote, lambda: node.publish(1))

	def testPublishCachedRemoteRetries(self) -> None:
		self.calledCounter = 0
		self.callbackRaise = True

		def callback(method: str, url: str, body: typing.Optional[bytes], **kwargs: typing.Any) -> typing.Any:
			self.calledCounter += 1
			if self.callbackRaise:
				raise Exception("Network is down")
			return "ok"

		node = Node(uid="testuid", remotes=["http://remote1"], httpClient=HttpClientMock(callback=callback))

		# First publish succeeds, caching the remote.
		self.callbackRaise = False
		node.publish(1)
		self.assertEqual(self.calledCounter, 1)

		# The cached remote is now retried 3 times before trying the sources.
		self.callbackRaise = True
		self.assertRaises(NodePublishNoRemote, lambda: node.publish(2))
		self.assertEqual(self.calledCounter, 5)

	def testPublishBlockForSRetries(self) -> None:
		self.calledCounter = 0

		def callback(method: str, url: str, body: typing.Optional[bytes], **kwargs: typing.Any) -> None:
			self.calledCounter += 1
			raise Exception("Unreachable")

		node = Node(uid="testuid", remotes=["http://remote1"], httpClient=HttpClientMock(callback=callback), blockForS=0.05)
		self.assertRaises(NodePublishNoRemote, lambda: node.publish(1))
		self.assertGreaterEqual(self.calledCounter, 2)

	def testExportGetErrorContinues(self) -> None:

		def callback(method: str, url: str, body: typing.Optional[bytes], **kwargs: typing.Any) -> typing.Any:
			if url.startswith("http://remote1"):
				raise Exception("Unreachable")
			if method == "GET":
				return HttpResponseMock(status=200, content=b'{"key": "value"}', headers={})
			raise Exception("Unexpected method")

		node = Node(uid="testuid", remotes=["http://remote1", "http://remote2"], httpClient=HttpClientMock(callback=callback))

		data = node.get(path=["a"])
		self.assertEqual(data, {"key": "value"})

		content = node.export(path=["a"])
		self.assertEqual(content, b'{"key": "value"}')

	def testPublishBulkEmpty(self) -> None:
		self.calledCounter = 0

		def callback(**kwargs: typing.Any) -> None:
			self.calledCounter += 1

		node = Node(uid="testuid", httpClient=HttpClientMock(callback=callback))
		with node.publishBulk() as _:
			pass
		self.assertEqual(self.calledCounter, 0)

	def testPublishMultiNodesEmpty(self) -> None:
		self.calledCounter = 0

		def callback(**kwargs: typing.Any) -> None:
			self.calledCounter += 1

		node = Node(uid="testuid", httpClient=HttpClientMock(callback=callback))
		node.publishMultiNodes(data={})
		self.assertEqual(self.calledCounter, 0)


if __name__ == "__main__":
	unittest.main()

File: apps/artifacts/api/nodejs/node/tests/node_test.js
import ExceptionFactory from "#bzd/nodejs/core/exception.js";
import { Node } from "#bzd/apps/artifacts/api/nodejs/node/node.js";

const Exception = ExceptionFactory("test", "artifacts", "node");

/// Build a Node instance with a mocked HTTP client, isolating tests from network and env.
function makeNode({ get = null, post = null, path = null } = {}) {
	const httpClient = {
		get: get ?? (async () => ({})),
		post: post ?? (async () => ({})),
	};
	return new Node({
		remotes: ["http://test"],
		volume: "nodes",
		token: null,
		httpClient,
		path,
	});
}

describe("Node", () => {
	describe("toTree", () => {
		it("Empty input returns empty object", () => {
			Exception.assertEqual(Node.toTree([]), {});
		});

		it("Single shallow entry", () => {
			Exception.assertEqual(Node.toTree([[["foo"], 42]]), { foo: 42 });
		});

		it("Nested key builds nested object", () => {
			Exception.assertEqual(Node.toTree([[["a", "b", "c"], 1]]), { a: { b: { c: 1 } } });
		});

		it("Same key overwrites keeping latest", () => {
			Exception.assertEqual(
				Node.toTree([
					[["k"], 1],
					[["k"], 2],
				]),
				{ k: 2 },
			);
		});

		it("Mix of paths merges into a shared tree", () => {
			Exception.assertEqual(
				Node.toTree([
					[["a", "x"], 1],
					[["a", "y"], 2],
					[["b"], 3],
				]),
				{ a: { x: 1, y: 2 }, b: 3 },
			);
		});
	});

	describe("get", () => {
		it("Returns the remote result unchanged when metadata is false", async () => {
			const canned = { data: [["k", "v"]], timestamp: 123 };
			const node = makeNode({ get: async () => canned });
			const result = await node.get({ remote: "http://test", uid: "u1" });
			Exception.assertEqual(result, canned);
		});

		it("Preserves the data array shape when metadata is true", async () => {
			const canned = {
				timestamp: 1000,
				data: [["k", [10, "v"]]],
			};
			const node = makeNode({ get: async () => canned });
			const result = await node.get({ remote: "http://test", uid: "u1", metadata: true });

			// Structural-only assertions: the metadata branch reshuffles the first
			// element of each value (t - timestampDiff) but must preserve the
			// outer array length, key, and inner value array length.
			Exception.assert(Array.isArray(result.data), "result.data must remain an array");
			Exception.assertEqual(result.data.length, 1);
			Exception.assertEqual(result.data[0][0], "k");
			Exception.assert(Array.isArray(result.data[0][1]), "inner value must remain an array");
			Exception.assertEqual(result.data[0][1].length, 2);
		});
	});

	describe("getAsTree", () => {
		it("Transforms flat result data into a tree", async () => {
			const canned = {
				timestamp: 0,
				data: [
					[["a", "x"], 1],
					[["a", "y"], 2],
					[["b"], 3],
				],
			};
			const node = makeNode({ get: async () => canned });
			const result = await node.getAsTree({ remote: "http://test", uid: "u1" });
			Exception.assertEqual(result.data, { a: { x: 1, y: 2 }, b: 3 });
		});
	});

	describe("list", () => {
		it("Collects node names from a single page", async () => {
			const node = makeNode({
				post: async () => ({
					data: { n1: { name: "n1" }, n2: { name: "n2" } },
					next: null,
				}),
			});
			const nodes = await node.list({ remote: "http://test" });
			Exception.assertEqual(nodes, ["n1", "n2"]);
		});
	});

	describe("publish", () => {
		it("Posts data with a client timestamp to the data endpoint", async () => {
			const calls = [];
			const node = makeNode({
				post: async (url, options) => {
					calls.push({ url, json: options.json });
					return {};
				},
			});
			await node.publish({ data: "hello", uid: "testuid" });

			Exception.assertEqual(calls.length, 1);
			Exception.assertEqual(calls[0].url, "http://test/x/nodes/testuid/data/");

			const body = calls[0].json;
			Exception.assert(Array.isArray(body.data));
			Exception.assertEqual(body.data[0][1][0][1], "hello");
			Exception.assert(typeof body.timestamp === "number");
		});

		it("Builds a nested URI from the publish path option", async () => {
			const urls = [];
			const node = makeNode({
				post: async (url) => {
					urls.push(url);
					return {};
				},
			});
			await node.publish({ data: 1, uid: "u", path: ["foo", "bar"] });
			Exception.assertEqual(urls[0], "http://test/x/nodes/u/data/foo/bar/");
		});

		it("Prepends the constructor path to the publish path", async () => {
			const urls = [];
			const node = makeNode({
				path: ["root"],
				post: async (url) => {
					urls.push(url);
					return {};
				},
			});
			await node.publish({ data: 1, uid: "u", path: ["foo"] });
			Exception.assertEqual(urls[0], "http://test/x/nodes/u/data/root/foo/");
		});

		it("Throws when posting to all remotes fails", async () => {
			const node = makeNode({
				post: async () => {
					throw new Error("Unreachable");
				},
			});
			await Exception.assertThrows(async () => {
				await node.publish({ data: 1, uid: "testuid" });
			});
		});
	});

	describe("publishBulk", () => {
		it("Posts accumulated entries from the callback to the bulk data endpoint", async () => {
			const calls = [];
			const node = makeNode({
				post: async (url, options) => {
					calls.push({ url, json: options.json });
					return {};
				},
			});
			await node.publishBulk({ uid: "testuid" }, (publish) => {
				publish({ timestampMs: 100, data: "a" });
				publish({ timestampMs: 200, data: "b" });
			});

			Exception.assertEqual(calls.length, 1);
			Exception.assertEqual(calls[0].url, "http://test/x/nodes/testuid/data/");
			Exception.assertEqual(calls[0].json.data, [
				[[], [[100, "a"]]],
				[[], [[200, "b"]]],
			]);
		});

		it("Builds a nested URI from the publishBulk path option", async () => {
			const urls = [];
			const node = makeNode({
				post: async (url) => {
					urls.push(url);
					return {};
				},
			});
			await node.publishBulk({ uid: "u", path: ["foo", "bar"] }, (publish) => {
				publish({ timestampMs: 1, data: "x" });
			});
			Exception.assertEqual(urls[0], "http://test/x/nodes/u/data/foo/bar/");
		});

		it("Prepends the constructor path to the publishBulk path", async () => {
			const urls = [];
			const node = makeNode({
				path: ["root"],
				post: async (url) => {
					urls.push(url);
					return {};
				},
			});
			await node.publishBulk({ uid: "u", path: ["foo"] }, (publish) => {
				publish({ timestampMs: 1, data: "x" });
			});
			Exception.assertEqual(urls[0], "http://test/x/nodes/u/data/root/foo/");
		});

		it("Throws and does not publish when the callback throws", async () => {
			const calls = [];
			const node = makeNode({
				post: async () => {
					calls.push(1);
					return {};
				},
			});
			await Exception.assertThrows(async () => {
				await node.publishBulk({ uid: "u" }, () => {
					throw new Error("boom");
				});
			});
			Exception.assertEqual(calls.length, 0);
		});

		it("Forwards isClientTimestamp false to the underlying _publish call", async () => {
			const calls = [];
			const node = makeNode({
				post: async (url, options) => {
					calls.push({ url, json: options.json });
					return {};
				},
			});
			await node.publishBulk({ uid: "u", isClientTimestamp: false }, (publish) => {
				publish({ timestampMs: 100, data: "x" });
			});

			Exception.assertEqual(calls[0].json.timestamp, undefined);
		});

		it("Forwards default isClientTimestamp true to the underlying _publish call", async () => {
			const calls = [];
			const node = makeNode({
				post: async (url, options) => {
					calls.push({ url, json: options.json });
					return {};
				},
			});
			await node.publishBulk({ uid: "u" }, (publish) => {
				publish({ timestampMs: 100, data: "x" });
			});

			Exception.assert(typeof calls[0].json.timestamp === "number");
		});

		it("Skips the POST when the bulk callback publishes nothing", async () => {
			const calls = [];
			const node = makeNode({
				post: async () => {
					calls.push(1);
					return {};
				},
			});
			await node.publishBulk({ uid: "testuid" }, () => {});
			Exception.assertEqual(calls.length, 0);
		});
	});
});
